// ngx-base-state/esm2020/lib/states/base.state.mjs
import { __decorate } from "tslib";
import { Inject, inject, Injectable, InjectionToken, Optional } from '@angular/core';
import { BehaviorSubject } from 'rxjs';
import { ɵNGX_STATE_DECORATOR_METADATA_FIELD } from '../constants';
import { ɵAction as Action } from '../decorators';
import { ɵMetadataKeyEnum, ɵMetadataOperationTypeEnum } from '../enums';
import { ɵMetadataStorage, ɵStackTrace } from '../helpers';
import { NGX_BASE_STATE_DEVTOOLS_CONFIG } from '../tokens';
import * as i0 from "@angular/core";
const INITIAL_DATA = new InjectionToken('__NGX_BASE_STATE_INITIAL_DATA');
const INITIAL_CONFIG = new InjectionToken('__NGX_BASE_STATE_INITIAL_CONFIG');
const CLASS_ID_FIELD = '_ɵID';
/**
 *	@class
 *	@classdes This is a base class that used for creating hight level state classes
 */
export class BaseState {
    constructor(
    /** Initial data should be passed via the `super` method call. */
    initialData = null, 
    /** Initial config should be passed via the `super` method call. */
    initialConfig = null) {
        this.initialData = initialData;
        this.initialConfig = initialConfig;
        this._devtoolsConfig = inject(NGX_BASE_STATE_DEVTOOLS_CONFIG);
        this._metadataStorage = inject(ɵMetadataStorage);
        this._currentlyInvokedAction = null;
        this._stackTraceOfCurrentlyInvokedAction = null;
        this._data$ = new BehaviorSubject(this.initialData);
        this.initClassIdIfAbsent();
        this.showConsoleWarningIfClassHaveNotDecorator();
        this.init();
    }
    /**
     * 	Get `Observable` with state data.
     *  @public
     *	@return {Generic} Observable with the state data.
     */
    get data$() {
        return this._data$.asObservable();
    }
    /**
     * 	Get state data.
     *  @public
     *	@return {Generic} State data.
     */
    get data() {
        return this._data$.value;
    }
    /**
     *  Base implementation of `ngOnDestroy`.
     *  Don't forget to call `super.ngOnDestroy` in case of override.
     *  @public
     */
    ngOnDestroy() {
        this.emitMetadataOperation(ɵMetadataOperationTypeEnum.Destroy);
    }
    /**
     *  Set new value to state
     *  @public
     *	@param {Generic} value - the value that should be set to update `BehaviorSubject`.
     */
    set(value) {
        this.setNewValue(value);
    }
    /**
     *  Clear state value. (Will be set `null`)
     *  @public
     */
    clear() {
        this.setNewValue(null);
    }
    /**
     *  Restore initial data from constructor.
     *  @public
     */
    restoreInitialData() {
        this.setNewValue(this.initialData);
    }
    /**
     *  Method for set data functionality. It may be expanded.
     *  The idea is to process the creation of new instances of complex structures.
     *  @protected
     *	@param {Generic | null} value - the value that should be set to update `BehaviorSubject`.
     */
    setNewValue(value) {
        this._data$.next(value);
        this.emitMetadataOperation(ɵMetadataOperationTypeEnum.Update);
    }
    /**
     * 	Method used for try to work out any method
     *  @protected
     *	@param {string} actionName - Action you try to fire. Used to show in Error text when something went wrong.
     *	@param {Function} actionFunc - Callback with logic. When something goes wrong - Error will be created.
     *	@return {Generic} result of the callback call.
     */
    tryDoAction(actionName, actionFunc) {
        try {
            return actionFunc();
        }
        catch (error) {
            this.catchError(error, actionName);
            return undefined;
        }
    }
    /**
     *  Method that	processed error for user friendly error messages
     *  @protected
     *	@param {Error} error - Error.
     *	@param {string} actionName - Name of the action where error happened.
     */
    catchError(error, actionName) {
        throw new Error(`\n${this.constructor.name} [${actionName}]: ${error.message}`);
    }
    init() {
        this.emitMetadataOperation(ɵMetadataOperationTypeEnum.Init);
    }
    // Using by decorators
    _onActionLikeInvoked(actionName) {
        if (this._devtoolsConfig.isEnabled && !this._currentlyInvokedAction) {
            this._currentlyInvokedAction = actionName;
            this._stackTraceOfCurrentlyInvokedAction = ɵStackTrace.capture();
        }
    }
    // Using by decorators
    _onActionLikeInvokeEnd() {
        if (this._devtoolsConfig.isEnabled) {
            this._currentlyInvokedAction = null;
            this._stackTraceOfCurrentlyInvokedAction = null;
        }
    }
    showConsoleWarningIfClassHaveNotDecorator() {
        if (this._devtoolsConfig.isEnabled && !this[ɵNGX_STATE_DECORATOR_METADATA_FIELD]) {
            console.warn(`${this.constructor.name} class is missed @NgxState() decorator. ` +
                `Some features of DevTools will work incorrectly!`);
        }
    }
    initClassIdIfAbsent() {
        if (!this.constructor[CLASS_ID_FIELD]) {
            this.constructor[CLASS_ID_FIELD] = Math.random();
        }
    }
    /**
     *  Emits information about state changes into `ReplaySubject` at the `window`.
     *  Extension use this information to visually represent current state and history of states changes.
     *  @private
     */
    emitMetadataOperation(type) {
        if (this._devtoolsConfig.isEnabled) {
            const operationEmitter$ = this._metadataStorage
                .get(ɵMetadataKeyEnum.MetadataOperation);
            operationEmitter$.next({
                type,
                classId: this.constructor[CLASS_ID_FIELD],
                className: this.constructor.name,
                classContext: this.initialConfig?.context,
                actionName: this._currentlyInvokedAction,
                date: new Date().toJSON(),
                data: this.data,
                stackTrace: this._stackTraceOfCurrentlyInvokedAction
            });
        }
    }
}
BaseState.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "14.2.4", ngImport: i0, type: BaseState, deps: [{ token: INITIAL_DATA, optional: true }, { token: INITIAL_CONFIG, optional: true }], target: i0.ɵɵFactoryTarget.Injectable });
BaseState.ɵprov = i0.ɵɵngDeclareInjectable({ minVersion: "12.0.0", version: "14.2.4", ngImport: i0, type: BaseState });
__decorate([
    Action
], BaseState.prototype, "set", null);
__decorate([
    Action
], BaseState.prototype, "clear", null);
__decorate([
    Action
], BaseState.prototype, "restoreInitialData", null);
__decorate([
    Action
], BaseState.prototype, "init", null);
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "14.2.4", ngImport: i0, type: BaseState, decorators: [{
            type: Injectable
        }], ctorParameters: function () { return [{ type: undefined, decorators: [{
                    type: Inject,
                    args: [INITIAL_DATA]
                }, {
                    type: Optional
                }] }, { type: undefined, decorators: [{
                    type: Inject,
                    args: [INITIAL_CONFIG]
                }, {
                    type: Optional
                }] }]; }, propDecorators: { set: [], clear: [], restoreInitialData: [], init: [] } });
