// ngx-base-state/esm2020/lib/states/record.state.mjs
import { __decorate } from "tslib";
import { map, shareReplay } from 'rxjs';
import { ɵAction as Action } from '../decorators';
import { BaseState } from './base.state';
/**
 *	@class
 *	@classdes Record state class. Used for save state with Record type.
 */
export class RecordState extends BaseState {
    constructor() {
        super(...arguments);
        /**
         *  Get `Observable` with all `keys` of your `Record` object state.
         *  @public
         */
        this.keys$ = this.data$
            .pipe(map(() => this.keys), shareReplay(1));
        /**
         *  Get `Observable` with all `values` of your `Record` object state.
         *  @public
         */
        this.values$ = this.data$
            .pipe(map(() => this.values), shareReplay(1));
    }
    /**
     *  Get all `keys` of your `Record` object state.
     *  @public
     */
    get keys() {
        return (this.data) ? Object.keys(this.data) : [];
    }
    /**
     *  Get all `values` of your `Record` object state.
     *  @public
     */
    get values() {
        return (this.data) ? Object.values(this.data) : [];
    }
    /**
     * 	Set item to object in state.
     *	@public
     *	@param {TKey} key - Key to set into.
     *	@param {TValue} value - Value to set within `key`.
     */
    setItem(key, value) {
        this.data[key] = value;
        this.setNewValue(this.data);
    }
    /**
     * 	Remove item from object in state.
     *	@public
     *	@param {TKey} key - Key to remove item within.
     */
    removeItem(key) {
        delete this.data[key];
        this.setNewValue(this.data);
    }
    /**
     * 	Remove all items from object in state.
     *	@public
     */
    removeAllItems() {
        this.set({});
    }
    setNewValue(value) {
        if (value) {
            super.setNewValue({ ...value });
        }
        else {
            super.setNewValue(null);
        }
    }
    catchError(error, actionName) {
        if (error instanceof TypeError) {
            throw new Error(`\n${this.constructor.name} [${actionName}]: ` +
                `Firstly set Object [Record].\n\n${error.message}`);
        }
        super.catchError(error, actionName);
    }
}
__decorate([
    Action
], RecordState.prototype, "setItem", null);
__decorate([
    Action
], RecordState.prototype, "removeItem", null);
__decorate([
    Action
], RecordState.prototype, "removeAllItems", null);
